src/sentry/eventstream/kafka/consumer_strategy.py

```diff
@@ -0,0 +1,146 @@
+import logging
+import time
+from collections import deque
+from concurrent.futures import Future, ThreadPoolExecutor
+from typing import Any, Deque, Mapping, Optional, Tuple
+
+from arroyo.backends.kafka.consumer import KafkaPayload
+from arroyo.processing.strategies import ProcessingStrategy, ProcessingStrategyFactory
+from arroyo.processing.strategies.abstract import MessageRejected
+from arroyo.types import Commit, Message, Partition, Position
+
+from sentry import options
+from sentry.eventstream.kafka.postprocessworker import _sampled_eventstream_timer
+from sentry.eventstream.kafka.postprocessworker import (
+    dispatch_post_process_group_task as _dispatch_post_process_group_task,
+)
+from sentry.eventstream.kafka.protocol import (
+    get_task_kwargs_for_message,
+    get_task_kwargs_for_message_from_headers,
+)
+from sentry.utils import metrics
+
+_DURATION_METRIC = "eventstream.duration"
+
+logger = logging.getLogger(__name__)
+
+
+# For testing. Function will eventually move here when postprocessworker is removed.
+def dispatch_post_process_group_task(
+    event_id: str,
+    project_id: int,
+    group_id: Optional[int],
+    is_new: bool,
+    is_regression: bool,
+    is_new_group_environment: bool,
+    primary_hash: Optional[str],
+    skip_consume: bool = False,
+) -> None:
+    _dispatch_post_process_group_task(
+        event_id,
+        project_id,
+        group_id,
+        is_new,
+        is_regression,
+        is_new_group_environment,
+        primary_hash,
+        skip_consume,
+    )
+
+
+def _get_task_kwargs(message: Message[KafkaPayload]) -> Optional[Mapping[str, Any]]:
+    use_kafka_headers = options.get("post-process-forwarder:kafka-headers")
+
+    if use_kafka_headers:
+        try:
+            with _sampled_eventstream_timer(instance="get_task_kwargs_for_message_from_headers"):
+                return get_task_kwargs_for_message_from_headers(message.payload.headers)
+        except Exception as error:
+            logger.warning("Could not forward message: %s", error, exc_info=True)
+            with metrics.timer(_DURATION_METRIC, instance="get_task_kwargs_for_message"):
+                return get_task_kwargs_for_message(message.payload.value)
+    else:
+        with metrics.timer(_DURATION_METRIC, instance="get_task_kwargs_for_message"):
+            return get_task_kwargs_for_message(message.payload.value)
+
+
+def _get_task_kwargs_and_dispatch(message: Message[KafkaPayload]) -> None:
+    task_kwargs = _get_task_kwargs(message)
+    if not task_kwargs:
+        return None
+
+    dispatch_post_process_group_task(**task_kwargs)
+
+
+class DispatchTask(ProcessingStrategy[KafkaPayload]):
```

Comment:
Will eventually be replaced by Arroyo's RunTask when that is available https://github.com/getsentry/arroyo/pull/111